### PR TITLE
manager: Integrated e2e test

### PR DESCRIPTION
**Pull request by froesef**
This PR will integrate the existing e2e test into the existing go suite for the manager that can be run as a unit test or as an integration test against a K8s environment.

### CODE REVIEW

File: manager/testdata/e2e/module-read.yaml [hunk 1]
```diff
@@ -5,11 +5,12 @@ apiVersion: app.m4d.ibm.com/v1alpha1
 kind: M4DModule
 metadata:
   name: arrow-flight-module
+  namespace: default
   labels:
     name: arrow-flight-module
     version: 0.0.1  # semantic version
 spec:
-  chart: ghcr.io/the-mesh-for-data/arrow-flight-module-chart:latest
+  chart: kind-registry:5000/m4d-system/arrow-flight-module-chart:latest
```
Comment (roee88) [inline manager/testdata/e2e/module-read.yaml:13]: quick question, why do we have to put it in the local registry?

File: manager/controllers/app/m4dapplication_controller_test.go [hunk 2]
```diff
@@ -406,4 +410,127 @@ var _ = Describe("M4DApplication Controller", func() {
...
+
+			// delete storage
+			_ = k8sClient.Delete(context.Background(), &apiv1alpha1.M4DBucket{ObjectMeta: metav1.ObjectMeta{Name: "test-bucket", Namespace: "default"}})
+			// delete modules
+			_ = k8sClient.Delete(context.Background(), &apiv1alpha1.M4DModule{ObjectMeta: metav1.ObjectMeta{Name: "arrow-flight-module", Namespace: "default"}})
+			_ = k8sClient.Delete(context.Background(), &apiv1alpha1.M4DModule{ObjectMeta: metav1.ObjectMeta{Name: "template-module", Namespace: "default"}})
+		})
+
+		It("Test end-to-end for M4DApplication with arrow-flight module", func() {
+			Expect(1).Should(Equal(1))
```
Comment (shlomitk1) [inline manager/controllers/app/m4dapplication_controller_test.go:435]: Why is this check needed?

File: manager/controllers/app/m4dapplication_controller_test.go [hunk 3]
```diff
@@ -406,4 +410,127 @@ var _ = Describe("M4DApplication Controller", func() {
...
+
+			var err error
+			bucketYAML, err := ioutil.ReadFile("../../testdata/e2e/bucket-available.yaml")
```
Comment (shlomitk1) [inline manager/controllers/app/m4dapplication_controller_test.go:438]: We don't need a bucket for read.

File: manager/controllers/app/m4dapplication_controller_test.go [hunk 4]
```diff
@@ -406,4 +410,127 @@ var _ = Describe("M4DApplication Controller", func() {
...
+			resource := &apiv1alpha1.M4DApplication{ObjectMeta: metav1.ObjectMeta{Name: "notebook", Namespace: "default"}}
+			_ = k8sClient.Delete(context.Background(), resource)
+
+			Eventually(func() error {
+				f := &apiv1alpha1.M4DApplication{}
+				return k8sClient.Get(context.Background(), appSignature, f)
+			}, timeout, interval).ShouldNot(Succeed())
+
+			// delete storage
+			_ = k8sClient.Delete(context.Background(), &apiv1alpha1.M4DBucket{ObjectMeta: metav1.ObjectMeta{Name: "test-bucket", Namespace: "default"}})
```
Comment (shlomitk1) [inline manager/controllers/app/m4dapplication_controller_test.go:428]: Should not be allocated.

File: manager/controllers/app/m4dapplication_controller_test.go [hunk 5]
```diff
@@ -406,4 +410,127 @@ var _ = Describe("M4DApplication Controller", func() {
...
+			// delete modules
+			_ = k8sClient.Delete(context.Background(), &apiv1alpha1.M4DModule{ObjectMeta: metav1.ObjectMeta{Name: "arrow-flight-module", Namespace: "default"}})
+			_ = k8sClient.Delete(context.Background(), &apiv1alpha1.M4DModule{ObjectMeta: metav1.ObjectMeta{Name: "template-module", Namespace: "default"}})
```
Comment (shlomitk1) [inline manager/controllers/app/m4dapplication_controller_test.go:431]: Why is template-module needed?

File: manager/testdata/e2e/module-read.yaml (hunk 1 above)
Comment (froesef) [reply manager/testdata/e2e/module-read.yaml:13]: For some reason the manager did not trust ghcr.io ... I did not check in depth why it was signed by an unknown authority but just uploaded it to the local registry that is trusted to make some progress. But you're right we should probably tackle that problem.

File: manager/controllers/app/m4dapplication_controller_test.go [hunk 6]
```diff
@@ -406,4 +410,109 @@ var _ = Describe("M4DApplication Controller", func() {
...
+
+			moduleKey, err := client.ObjectKeyFromObject(module)
+			Expect(err).ToNot(HaveOccurred())
+
+			// Create M4DModule
+			Expect(k8sClient.Create(context.Background(), module)).Should(Succeed())
+
+			// Ensure getting cleaned up after tests finish
+			defer func() {
+				f := &apiv1alpha1.M4DBucket{ObjectMeta: metav1.ObjectMeta{Namespace: moduleKey.Namespace, Name: moduleKey.Name}}
```
Comment (shlomitk1) [inline manager/controllers/app/m4dapplication_controller_test.go:447]: Did you mean M4DModule?

File: manager/controllers/app/m4dapplication_controller_test.go [hunk 7]
```diff
@@ -406,4 +410,109 @@ var _ = Describe("M4DApplication Controller", func() {
...
+
+			applicationKey, err := client.ObjectKeyFromObject(application)
+			Expect(err).ToNot(HaveOccurred())
+
+			// Create M4DApplication
+			Expect(k8sClient.Create(context.Background(), application)).Should(Succeed())
+
+			// Ensure getting cleaned up after tests finish
+			defer func() {
+				f := &apiv1alpha1.M4DBucket{ObjectMeta: metav1.ObjectMeta{Namespace: applicationKey.Namespace, Name: applicationKey.Name}}
```
Comment (shlomitk1) [inline manager/controllers/app/m4dapplication_controller_test.go:466]: M4DApplication?

File: manager/controllers/app/suite_test.go [hunk 8]
```diff
@@ -47,21 +51,23 @@ var _ = BeforeSuite(func(done Done) {
 	logf.SetLogger(zap.LoggerTo(GinkgoWriter, true))
 
 	By("bootstrapping test environment")
+	if os.Getenv("NO_SIMULATED_PROGRESS") == "true" {
+		noSimulatedProgress = true
```
Comment (shlomitk1) [inline manager/controllers/app/suite_test.go:55]: What does noSimulatedProgress test?

File: manager/controllers/utils/configuration.go [hunk 9]
```diff
@@ -107,3 +107,30 @@ func GetCredentialsManagerServiceAddress() string {
...
+	SetIfNotSet(VaultDatasetHomeKey, "m4d/dataset-creds/", t)
+	SetIfNotSet(VaultUserHomeKey, "m4d/user-creds/", t)
+	SetIfNotSet(VaultTTLKey, "24h", t)
+	SetIfNotSet(VaultAuthKey, "kubernetes", t)
+	SetIfNotSet("RUN_WITHOUT_VAULT", "1", t)
+	SetIfNotSet("ENABLE_WEBHOOKS", "false", t)
+	SetIfNotSet("CONNECTION_TIMEOUT", "120", t)
+	SetIfNotSet("MAIN_POLICY_MANAGER_CONNECTOR_URL", "localhost:50090", t)
+	SetIfNotSet("MAIN_POLICY_MANAGER_NAME", "MOCK", t)
+	SetIfNotSet("USE_EXTENSIONPOLICY_MANAGER", "http://secret-provider.m4d-system.svc.cluster.local:5555/get-secret", t)
```
Comment (shlomitk1) [inline manager/controllers/utils/configuration.go:134]: Mix of extensionpolicy (true/false) with SecretProviderURL

File: manager/testdata/e2e/bucket-available.yaml [hunk 10]
```diff
@@ -5,6 +5,7 @@ apiVersion: app.m4d.ibm.com/v1alpha1
 kind: M4DBucket
 metadata:
   name: test-bucket
+  namespace: default
```
Comment (shlomitk1) [inline manager/testdata/e2e/bucket-available.yaml:8]: Should run in m4d-system

File: manager/testdata/e2e/module-read.yaml [hunk 11]
```diff
@@ -5,11 +5,12 @@ apiVersion: app.m4d.ibm.com/v1alpha1
 kind: M4DModule
 metadata:
   name: arrow-flight-module
+  namespace: default
```
Comment (shlomitk1) [inline manager/testdata/e2e/module-read.yaml:8]: m4d-system

File: manager/testdata/e2e/module-template.yaml [hunk 12]
```diff
@@ -0,0 +1,33 @@
+# Copyright 2020 IBM Corp.
```
Comment (shlomitk1) [inline manager/testdata/e2e/module-template.yaml:1]: module-template is not required. I think it was left as a template for how the modules should be written (charts, templates, values,...)

File: modules/m4d-template/templates/configmap.yaml [hunk 13]
```diff
@@ -0,0 +1,14 @@
+# Copyright 2020 IBM Corp.
```
Comment (shlomitk1) [inline modules/m4d-template/templates/configmap.yaml:1]: Why is this file required?

File: modules/m4d-template/values.yaml [hunk 14]
```diff
@@ -5,4 +5,9 @@
 # This is a YAML-formatted file.
 # Declare variables to be passed into your templates.
 
+metadata:
+  name: "hello-world"
+  namespace: "default"
```
Comment (shlomitk1) [inline modules/m4d-template/values.yaml:10]: Namespace should not be defined - it is taken from the blueprint during orchestration.

File: manager/testdata/e2e/module-read.yaml (hunk 1 above)
Comment (froesef) [reply manager/testdata/e2e/module-read.yaml]: Ok. I fixed the certificate issue!

File: manager/Makefile [hunk 15]
```diff
@@ -56,53 +56,20 @@ run-integration-tests: wait_for_manager
...
-	$(ABSTOOLBIN)/kubectl delete -f testdata/e2e/module-implicit-copy-kafka-to-s3-stream.yaml || true
-	$(ABSTOOLBIN)/kubectl patch m4dapplication notebook -p '{"metadata":{"finalizers":[]}}' --type=merge || true
-	$(ABSTOOLBIN)/kubectl delete -f testdata/e2e/m4dapplication.yaml -n default || true
-	@echo
-	$(MAKE) uninstall 1>/dev/null 2>/dev/null || true
+# Download CA certificates from manager image and add the CA certificate that was generated for the test
+# This file is then uploaded to K8s as a secret and injected into the manager during deployment
+.PHONY: build-ca-certs
+build-ca-certs: $(TOOLBIN)/kubectl
+	rm ../hack/registry/ca-certificates.crt || true
```
Comment (roee88) [inline manager/Makefile:66]: I think we usually use ROOT_DIR rather than relative paths

File: .gitignore [hunk 16]
```diff
@@ -49,3 +49,6 @@ test/services/pilot/policy-manager-mock/policy-manager-mock
 connectors/opa/opa_connector
 connectors/vault/vault_connector
 connectors/egeria/target
+
+# Downloaded arrow-flight-module
+modules/arrow-flight-module
```
Comment (roee88) [inline .gitignore:54]: Why are they downloaded? and where

File: manager/controllers/app/suite_test.go [hunk 17]
```diff
@@ -51,17 +57,16 @@ var _ = BeforeSuite(func(done Done) {
 		CRDDirectoryPaths: []string{
 			filepath.Join("..", "..", "config", "crd", "bases"),
 		},
-		AttachControlPlaneOutput: true,
+		//AttachControlPlaneOutput: true,
```
Comment (roee88) [inline manager/controllers/app/suite_test.go:60]: Remove if not needed

File: manager/testdata/e2e/bucket-available.yaml (hunk 10 above)
Comment (roee88) [inline manager/testdata/e2e/bucket-available.yaml:8]: why is @shlomitk1's comment marked resolved? I have the same comment

**Review by roee88 — COMMENTED** [review]
Can you please explain in a comment what are the effects on the certificates in the manager image?

**Review by roee88 — APPROVED** [review]
